pennylane/devices/qutrit_mixed/measure.py

```diff
@@ -0,0 +1,301 @@
+
+    Returns:
+        TensorLike: the result of obs@state.
+    """
+
+    num_ch_wires = len(obs.wires)
+    einsum_indices = get_einsum_mapping(obs, state, _map_indices_apply_operation, is_state_batched)
+    obs_mat = obs.matrix()
+    obs_shape = [QUDIT_DIM] * num_ch_wires * 2
+    obs_mat = math.cast(math.reshape(obs_mat, obs_shape), complex)
+    return math.einsum(einsum_indices, obs_mat, state)
+
+
+def calculate_expval(
+    measurementprocess: ExpectationMP, state: TensorLike, is_state_batched: bool = False
+) -> TensorLike:
+    """Measure the expectation value of an observable by finding the trace of obs@rho.
+
+    Args:
+        measurementprocess (ExpectationMP): measurement process to apply to the state.
+        state (TensorLike): the state to measure.
+        is_state_batched (bool): whether the state is batched or not.
+
+    Returns:
+        TensorLike: the result of the measurement.
+    """
+    obs = measurementprocess.obs
+    rho_mult_obs = apply_observable_einsum(obs, state, is_state_batched)
+
+    # using einsum since trace function axis selection parameter names
+    # are not consistent across interfaces, they don't exist for torch
+
+    num_wires = get_num_wires(state, is_state_batched)
+    rho_mult_obs_reshaped = reshape_state_as_matrix(rho_mult_obs, num_wires)
+    if is_state_batched:
+        return math.real(math.stack([math.sum(math.diagonal(dm)) for dm in rho_mult_obs_reshaped]))
+
+    return math.real(math.sum(math.diagonal(rho_mult_obs_reshaped)))
+
+
+def calculate_reduced_density_matrix(  # TODO: ask if I should have state diagonalization gates?
+    measurementprocess: StateMeasurement, state: TensorLike, is_state_batched: bool = False
+) -> TensorLike:
+    """Get the state or reduced density matrix.
+
+    Args:
+        measurementprocess (StateMeasurement): measurement to apply to the state.
+        state (TensorLike): state to apply the measurement to.
+        is_state_batched (bool): whether the state is batched or not.
+
+    Returns:
+        TensorLike: the resulting resquared density_matrix.
+    """
+    wires = measurementprocess.wires
+    if not wires:
+        return reshape_state_as_matrix(state, get_num_wires(state, is_state_batched))
+
+    num_obs_wires = len(wires)
+    num_state_wires = get_num_wires(state, is_state_batched)
+    state_wire_indices_list = list(alphabet[:num_state_wires] * 2)
+    final_state_wire_indices_list = [""] * (2 * num_obs_wires)
+
+    for i, wire in enumerate(wires):
+        col_index = wire + num_state_wires
+        state_wire_indices_list[col_index] = alphabet[col_index]
+        final_state_wire_indices_list[i] = alphabet[wire]
+        final_state_wire_indices_list[i + num_obs_wires] = alphabet[col_index]
+
+    state_wire_indices = "".join(state_wire_indices_list)
+    final_state_wire_indices = "".join(final_state_wire_indices_list)
+
+    state = math.einsum(f"...{state_wire_indices}->...{final_state_wire_indices}", state)
```

Comment:
I had to remove some einsums used for diagonalization and tracing because it breaks when trying to find the derivative. I did this in other places.

Here I left reduce dm as I assume it will be uncommon to take the derivative of a reduced density matrix (especially since it would have to be flattened to do so), I could fix it so that it works, but it would require that I copy a lot of code or abstract the base code to work with qutrits. I think it's best to leave as is for now

Reply:
Mind providing an example of it breaking derivatives? It seems like the ml frameworks should be really good at handling einsum, so I'm a little confused about what would be breaking.

Reply:
```
state = qml.numpy.array(state)

def f(x):
    new_state = x * state
    diag = qml.math.einsum("abcabc->abc", new_state)
    return qml.math.flatten(diag)

x = qml.numpy.array(0.2, requires_grad=True)

g = qml.jacobian(f)(x)
```

Returns error:
```
[/usr/local/lib/python3.10/dist-packages/numpy/core/einsumfunc.py](https://localhost:8080/#) in einsum(out, optimize, *operands, **kwargs)
   1369         if specified_out:
   1370             kwargs['out'] = out
-> 1371         return c_einsum(*operands, **kwargs)
   1372 
   1373     # Check the kwargs to avoid a more cryptic error later, without having to

ValueError: einstein sum subscripts string includes output subscript 'a' multiple times
```
The einsum subscripts are swapped during backpropogation so `"abcabc->abc"` => `"abc->abcabc"` can't have repeating values in RHS. The same error occurs with trace.